File: education/education/page/new_registered_students/new_registered_students.js
// Copyright (c) 2026, Frappe Technologies Pvt. Ltd. and contributors
// For license information, please see license.txt

frappe.pages['new-registered-students'].on_page_load = function (wrapper) {
	const page = frappe.ui.make_app_page({
		parent: wrapper,
		title: __('New Registered Students'),
		single_column: true,
	});

	const METHOD = 'education.education.page.new_registered_students.new_registered_students';

	frappe.call({
		method: METHOD + '.get_page_html',
		callback: function (r) {
			if (r.message) {
				page.main.html(r.message);
				new NewRegisteredStudents(page, METHOD);
			} else {
				page.main.html(
					'<div class="alert alert-danger">' + __('Failed to load the page.') + '</div>'
				);
			}
		},
	});
};

class NewRegisteredStudents {
	constructor(page, method) {
		this.page = page;
		this.method = method;
		this.filters = {};
		this.page_length = 50;
		this.page_start = 0;
		this.total_count = 0;
		this.deleting = false;

		this.setup_program_filter();
		this.bind_events();
		this.bind_realtime();
		this.load();
	}

	// ---- helpers -------------------------------------------------------
	$(sel) {
		return this.page.main.find(sel);
	}

	setup_program_filter() {
		this.program_field = frappe.ui.form.make_control({
			parent: this.$('#nrs-program'),
			df: {
				fieldtype: 'Link',
				fieldname: 'program',
				options: 'Program',
				placeholder: __('Select Program'),
			},
			render_input: true,
		});
		this.program_field.refresh();
		// Apply filters when a program is picked
		this.program_field.$input.on('change', () => this.$('#nrs-apply').click());
	}

	collect_filters() {
		const f = {};
		const search = this.$('#nrs-search').val();
		const year = this.$('#nrs-educational-year').val();
		const program = this.program_field ? this.program_field.get_value() : '';
		const from_date = this.$('#nrs-from-date').val();
		const to_date = this.$('#nrs-to-date').val();
		if (search) f.search = search;
		if (year) f.educational_year = year;
		if (program) f.program = program;
		if (from_date) f.from_date = from_date;
		if (to_date) f.to_date = to_date;
		return f;
	}

	// ---- events --------------------------------------------------------
	bind_events() {
		this.$('#nrs-apply').on('click', () => {
			this.filters = this.collect_filters();
			this.page_start = 0;
			this.load();
		});

		this.$('#nrs-clear').on('click', () => {
			this.$('#nrs-search').val('');
			this.$('#nrs-educational-year').val('');
			if (this.program_field) this.program_field.set_value('');
			this.$('#nrs-from-date').val('');
			this.$('#nrs-to-date').val('');
			this.filters = {};
			this.page_start = 0;
			this.load();
		});

		this.$('#nrs-search, #nrs-educational-year').on('keypress', (e) => {
			if (e.which === 13) this.$('#nrs-apply').click();
		});

		// Jump to a typed page number
		this.$('#nrs-page-number').on('keypress', (e) => {
			if (e.which === 13) this.goto_page();
		});
		this.$('#nrs-page-number').on('change', () => this.goto_page());

		this.$('#nrs-prev').on('click', () => {
			if (this.page_start > 0) {
				this.page_start = Math.max(0, this.page_start - this.page_length);
				this.load();
			}
		});

		this.$('#nrs-next').on('click', () => {
			if (this.page_start + this.page_length < this.total_count) {
				this.page_start += this.page_length;
				this.load();
			}
		});

		this.$('#nrs-export').on('click', () => this.export_excel());
		this.$('#nrs-delete-all').on('click', () => this.delete_all());

		// Per-row delete (delegated)
		this.page.main.on('click', '.nrs-delete-row', (e) => {
			const student = $(e.currentTarget).data('student');
			const name = $(e.currentTarget).data('name');
			this.delete_one(student, name);
		});
	}

	bind_realtime() {
		frappe.realtime.on('new_registered_students_delete_progress', (data) => {
			this.update_progress(data);
		});
		frappe.realtime.on('new_registered_students_delete_done', (data) => {
			this.finish_deletion(data);
		});
	}

	// ---- data load -----------------------------------------------------
	load() {
		const tbody = this.$('#nrs-table-body');
		tbody.html(
			'<tr><td colspan="8" class="text-center text-muted">' +
				__('Loading...') +
				'</td></tr>'
		);

		frappe.call({
			method: this.method + '.get_students',
			args: {
				filters: this.filters,
				page_length: this.page_length,
				page_start: this.page_start,
			},
			callback: (r) => {
				if (!r.message) return;
				this.total_count = r.message.total_count;
				this.render_rows(r.message.data);
				this.render_summary();
				if (r.message.delete_running && !this.deleting) {
					this.set_deleting(true);
					this.$('#nrs-progress-wrap').show();
					this.$('#nrs-progress-label').text(__('Deletion in progress...'));
				}
			},
		});
	}

	render_rows(rows) {
		const tbody = this.$('#nrs-table-body');
		tbody.empty();

		if (!rows || !rows.length) {
			tbody.html(
				'<tr><td colspan="8" class="text-center text-muted">' +
					__('No matching students found.') +
					'</td></tr>'
			);
			return;
		}

		rows.forEach((row) => {
			const tr = $('<tr>');
			tr.append($('<td>').html(this.student_link(row.student_id)));
			tr.append($('<td>').text(row.student_name || ''));
			tr.append($('<td>').text(row.full_name || ''));
			tr.append($('<td>').text(row.email || ''));
			tr.append($('<td>').text(row.mobile_no || ''));
			tr.append($('<td>').text(row.joining_date ? frappe.datetime.str_to_user(row.joining_date) : ''));
			tr.append($('<td>').text(row.educational_year || ''));

			const btn = $(
				'<button class="btn btn-xs btn-danger nrs-delete-row">' +
					'<i class="fa fa-trash"></i> ' +
					__('Delete') +
					'</button>'
			);
			btn.attr('data-student', row.student_id);
			btn.attr('data-name', row.student_name || row.student_id);
			if (this.deleting) btn.prop('disabled', true);
			tr.append($('<td class="text-center">').append(btn));
			tbody.append(tr);
		});
	}

	student_link(id) {
		return (
			'<a href="/app/student/' +
			encodeURIComponent(id) +
			'" target="_blank">' +
			frappe.utils.escape_html(id) +
			'</a>'
		);
	}

	render_summary() {
		this.$('#nrs-total-count').text(this.total_count);
		const from = this.total_count ? this.page_start + 1 : 0;
		const to = Math.min(this.page_start + this.page_length, this.total_count);
		this.$('#nrs-page-info').text(__('{0} - {1} of {2}', [from, to, this.total_count]));

		const total_pages = Math.max(1, Math.ceil(this.total_count / this.page_length));
		const current_page = Math.floor(this.page_start / this.page_length) + 1;
		this.$('#nrs-total-pages').text(total_pages);
		this.$('#nrs-page-number').val(current_page).attr('max', total_pages);

		this.$('#nrs-prev').prop('disabled', this.page_start === 0);
		this.$('#nrs-next').prop(
			'disabled',
			this.page_start + this.page_length >= this.total_count
		);
	}

	goto_page() {
		const total_pages = Math.max(1, Math.ceil(this.total_count / this.page_length));
		let page = cint(this.$('#nrs-page-number').val());
		if (!page || page < 1) page = 1;
		if (page > total_pages) page = total_pages;

		const new_start = (page - 1) * this.page_length;
		if (new_start === this.page_start) {
			// value unchanged (or clamped back) — just re-sync the input
			this.$('#nrs-page-number').val(page);
			return;
		}
		this.page_start = new_start;
		this.load();
	}

	// ---- single row deletion ------------------------------------------
	delete_one(student, name) {
		frappe.confirm(
			__('Permanently delete student <b>{0}</b> and all related records (user, program enrollment, fees, course enrollment applicant)?', [
				frappe.utils.escape_html(name),
			]),
			() => {
				frappe.dom.freeze(__('Deleting {0}...', [name]));
				frappe.call({
					method: this.method + '.delete_student',
					args: { student: student },
					always: () => frappe.dom.unfreeze(),
					callback: (r) => {
						const res = r.message || {};
						if (res.success) {
							frappe.show_alert({
								message: __('Deleted {0}', [name]),
								indicator: 'green',
							});
							this.load();
						} else {
							frappe.msgprint({
								title: __('Deletion Failed'),
								message: frappe.utils.escape_html(res.error || __('Unknown error')),
								indicator: 'red',
							});
						}
					},
				});
			}
		);
	}

	// ---- bulk deletion -------------------------------------------------
	delete_all() {
		if (this.deleting) return;
		if (!this.total_count) {
			frappe.msgprint(__('There are no matching students to delete.'));
			return;
		}

		frappe.confirm(
			__('This will permanently delete <b>{0}</b> students and all their related records (users, program enrollments, fees, course enrollment applicants). This cannot be undone. Continue?', [
				this.total_count,
			]),
			() => {
				frappe.call({
					method: this.method + '.delete_all',
					args: { filters: this.filters },
					callback: (r) => {
						const res = r.message || {};
						if (res.enqueued) {
							this.start_deletion(res.total);
						} else {
							frappe.msgprint(res.message || __('Could not start deletion.'));
						}
					},
				});
			}
		);
	}

	start_deletion(total) {
		this.set_deleting(true);
		this.$('#nrs-failures').hide();
		this.$('#nrs-failures-body').empty();
		this.$('#nrs-progress-wrap').show();
		this.$('#nrs-progress-label').text(__('Deleting students...'));
		this.$('#nrs-progress-counts').text(__('0 of {0}', [total]));
		this.update_bar(0);
		frappe.show_alert({
			message: __('Deletion started in the background.'),
			indicator: 'blue',
		});
	}

	update_progress(data) {
		if (!this.deleting) {
			// A job was already running when the page loaded
			this.set_deleting(true);
			this.$('#nrs-progress-wrap').show();
		}
		const pct = data.total ? Math.round((data.current / data.total) * 100) : 0;
		this.update_bar(pct);
		this.$('#nrs-progress-counts').text(
			__('{0} of {1} — {2} deleted, {3} failed', [
				data.current,
				data.total,
				data.deleted,
				data.failed,
			])
		);
	}

	update_bar(pct) {
		this.$('#nrs-progress-bar').css('width', pct + '%').text(pct + '%');
	}

	finish_deletion(data) {
		this.update_bar(100);
		this.set_deleting(false);
		this.$('#nrs-progress-wrap').hide();

		const failures = (data && data.failures) || [];
		if (failures.length) {
			this.render_failures(failures);
			frappe.msgprint({
				title: __('Deletion Completed with Errors'),
				message: __('{0} deleted, {1} failed. See the failure details below.', [
					data.deleted,
					failures.length,
				]),
				indicator: 'orange',
			});
		} else {
			frappe.show_alert({
				message: __('All {0} students deleted successfully.', [data.deleted]),
				indicator: 'green',
			});
		}
		this.page_start = 0;
		this.load();
	}

	render_failures(failures) {
		const body = this.$('#nrs-failures-body');
		body.empty();
		failures.forEach((f) => {
			const tr = $('<tr>');
			tr.append($('<td>').text(f.student));
			tr.append($('<td>').text(f.student_name || ''));
			tr.append($('<td>').text(f.error || ''));
			body.append(tr);
		});
		this.$('#nrs-failures-count').text(failures.length);
		this.$('#nrs-failures').show();
	}

	// ---- UI state ------------------------------------------------------
	set_deleting(state) {
		this.deleting = state;
		this.$('#nrs-delete-all').prop('disabled', state);
		this.$('#nrs-export').prop('disabled', state);
		this.page.main.find('.nrs-delete-row').prop('disabled', state);
	}

	// ---- export --------------------------------------------------------
	export_excel() {
		const args = $.param({ filters: JSON.stringify(this.filters) });
		const url = '/api/method/' + this.method + '.export_students?' + args;
		window.open(url, '_blank');
	}
}

File: education/education/page/new_registered_students/new_registered_students.py
# Copyright (c) 2026, Frappe Technologies Pvt. Ltd. and contributors
# For license information, please see license.txt

import json

import frappe
from frappe import _
from frappe.utils import cint, cstr, strip_html_tags

# Realtime event names (must match the frontend subscriptions)
PROGRESS_EVENT = "new_registered_students_delete_progress"
DONE_EVENT = "new_registered_students_delete_done"
# Cache key used to flag that a background "delete all" job is currently running
RUNNING_CACHE_KEY = "new_registered_students_delete_running"

# Doctypes that link to a Student and must be removed before the Student itself.
# Order matters: dependents first, Student last, User after the Student.
STUDENT_DEPENDENTS = [
	"Fees",
	#"Course Enrollment",
	"Course Enrollment Applicant",
	"Drop Course",
	"Drop Coursepack",
	"Study Postponement",
	"Program Enrollment",
]


def _check_permission():
	"""Only System Managers may use this page and its server methods."""
	if "System Manager" not in frappe.get_roles(frappe.session.user):
		frappe.throw(_("Not permitted"), frappe.PermissionError)


# ---------------------------------------------------------------------------
# Page rendering
# ---------------------------------------------------------------------------


@frappe.whitelist()
def get_page_html():
	"""Return the rendered HTML body of the page."""
	_check_permission()
	return frappe.render_template(
		"education/page/new_registered_students/new_registered_students.html", {}
	)


# ---------------------------------------------------------------------------
# Listing
# ---------------------------------------------------------------------------


def _parse_filters(filters):
	if isinstance(filters, str):
		filters = json.loads(filters or "{}")
	return filters or {}


def _build_conditions(filters):
	"""Build the WHERE conditions and parameters for the student query.

	A student qualifies as a deletable "new registered student" when they:
	  * are NOT a scholarship student (no Program Enrollment with has_scholarship),
	  * have NEVER made any payment (no Fees with a paid amount or an uploaded receipt),
	  * have NO course enrollment.
	"""
	conditions = [
		# Not a scholarship student
		"""s.name NOT IN (
			SELECT pe.student FROM `tabProgram Enrollment` pe
			WHERE pe.has_scholarship = 1 AND pe.student IS NOT NULL
		)""",
		# Never made any payment: no fee with money received or a receipt attached
		"""s.name NOT IN (
			SELECT f.student FROM `tabFees` f
			WHERE f.student IS NOT NULL
			AND (
				(f.grand_total - f.outstanding_amount) > 0
				OR IFNULL(f.receipt_amount, 0) > 0
				OR IFNULL(f.receipt_uploaded, 0) = 1
			)
		)""",
		# No course enrollment
		"""s.name NOT IN (
			SELECT ce.student FROM `tabCourse Enrollment` ce
			WHERE ce.student IS NOT NULL
		)""",
		# No Journal Entry Account
		"""s.name NOT IN (
			SELECT ja.party FROM `tabJournal Entry Account` ja
			WHERE ja.party IS NOT NULL
		)""",
	]
	values = {}

	if filters.get("educational_year"):
		conditions.append("s.educational_year = %(educational_year)s")
		values["educational_year"] = filters["educational_year"]

	if filters.get("program"):
		# A student's program lives in their Program Enrollment(s) (a student may
		# have several, or none). Match on the enrollment rather than the
		# unreliable Student.program field.
		conditions.append(
			"""s.name IN (
				SELECT pe.student FROM `tabProgram Enrollment` pe
				WHERE pe.program = %(program)s AND pe.student IS NOT NULL
			)"""
		)
		values["program"] = filters["program"]

	if filters.get("from_date"):
		conditions.append("s.joining_date >= %(from_date)s")
		values["from_date"] = filters["from_date"]

	if filters.get("to_date"):
		conditions.append("s.joining_date <= %(to_date)s")
		values["to_date"] = filters["to_date"]

	if filters.get("search"):
		conditions.append(
			"""(s.name LIKE %(search)s
				OR s.student_name LIKE %(search)s
				OR s.student_email_id LIKE %(search)s)"""
		)
		values["search"] = "%" + filters["search"] + "%"

	return " AND ".join(conditions), values


@frappe.whitelist()
def get_students(filters=None, page_length=50, page_start=0):
	"""Return the paginated list of deletable new registered students."""
	_check_permission()
	filters = _parse_filters(filters)
	where_clause, values = _build_conditions(filters)

	total_count = frappe.db.sql(
		f"SELECT COUNT(*) FROM `tabStudent` s WHERE {where_clause}", values
	)[0][0]

	values["page_length"] = cint(page_length)
	values["page_start"] = cint(page_start)

	data = frappe.db.sql(
		f"""
		SELECT
			s.name AS student_id,
			s.student_name,
			CONCAT_WS(' ', s.first_name, s.middle_name, s.last_name) AS full_name,
			s.student_email_id AS email,
			s.student_mobile_number AS mobile_no,
			s.joining_date,
			s.educational_year,
			s.user,
			s.creation
		FROM `tabStudent` s
		WHERE {where_clause}
		ORDER BY s.creation DESC
		LIMIT %(page_length)s OFFSET %(page_start)s
		""",
		values,
		as_dict=True,
	)

	return {
		"data": data,
		"total_count": total_count,
		"page_length": cint(page_length),
		"page_start": cint(page_start),
		"delete_running": bool(frappe.cache().get_value(RUNNING_CACHE_KEY)),
	}


# ---------------------------------------------------------------------------
# Deletion
# ---------------------------------------------------------------------------


def _delete_student_data(student_id):
	"""Delete a single student and everything linked to them.

	All work happens inside a savepoint so that a failure on any one student
	leaves that student fully intact (all-or-nothing per student).

	Returns a tuple ``(success: bool, error: str)``.
	"""
	savepoint = "del_" + frappe.generate_hash(length=8)
	frappe.db.savepoint(savepoint)
	try:
		# 1. Remove all dependent documents that link to the student
		for doctype in STUDENT_DEPENDENTS:
			if not frappe.db.exists("DocType", doctype):
				continue
			for name in frappe.get_all(doctype, filters={"student": student_id}, pluck="name"):
				doc = frappe.get_doc(doctype, name)
				if getattr(doc, "docstatus", 0) == 1:
					doc.cancel()
				frappe.delete_doc(doctype, name, ignore_permissions=True, delete_permanently=True)

		# 2. Capture the linked user before deleting the student
		user = frappe.db.get_value("Student", student_id, "user")

		# 3. Delete the student
		frappe.delete_doc("Student", student_id, ignore_permissions=True, delete_permanently=True)

		# 4. Delete the linked user (and its access logs first)
		if user and frappe.db.exists("User", user):
			for log in frappe.get_all("Access Log", filters={"user": user}, pluck="name"):
				frappe.delete_doc("Access Log", log, ignore_permissions=True, delete_permanently=True)
			# Notification Settings
			for nts in frappe.get_all("Notification Settings", filters={"user": user}, pluck="name"):
				print("Deleting Notification Settings", nts)
				frappe.delete_doc("Notification Settings", nts, ignore_permissions=True, delete_permanently=True)
			frappe.delete_doc("User", user, ignore_permissions=True, delete_permanently=True)

		return True, ""
	except Exception as e:
		# Undo every change made for this student so it is left untouched
		frappe.db.rollback(save_point=savepoint)
		frappe.log_error(frappe.get_traceback(), f"Delete New Registered Student: {student_id}")
		err = strip_html_tags(cstr(e)).strip() or e.__class__.__name__
		frappe.clear_messages()
		return False, err


@frappe.whitelist()
def delete_student(student):
	"""Delete a single student synchronously (used by the per-row delete button)."""
	_check_permission()
	if not frappe.db.exists("Student", student):
		return {"success": False, "error": _("Student {0} no longer exists").format(student)}

	student_name = frappe.db.get_value("Student", student, "student_name")
	success, error = _delete_student_data(student)
	if success:
		frappe.db.commit()
	return {
		"success": success,
		"error": error,
		"student": student,
		"student_name": student_name,
	}


@frappe.whitelist()
def delete_all(filters=None):
	"""Enqueue a background job that deletes every student matching the filters."""
	_check_permission()

	if frappe.cache().get_value(RUNNING_CACHE_KEY):
		return {"enqueued": False, "message": _("A deletion job is already running.")}

	filters = _parse_filters(filters)
	where_clause, values = _build_conditions(filters)
	student_ids = frappe.db.sql(
		f"SELECT s.name FROM `tabStudent` s WHERE {where_clause} ORDER BY s.creation DESC",
		values,
		pluck=True,
	)

	if not student_ids:
		return {"enqueued": False, "message": _("No matching students to delete.")}

	# Flag the job as running (1 hour ttl as a safety net)
	frappe.cache().set_value(RUNNING_CACHE_KEY, 1, expires_in_sec=3600)

	frappe.enqueue(
		"education.education.page.new_registered_students.new_registered_students.delete_all_background",
		queue="long",
		timeout=14400,
		student_ids=student_ids,
		user=frappe.session.user,
	)

	return {"enqueued": True, "total": len(student_ids)}


def delete_all_background(student_ids, user):
	"""Background worker: delete all students and stream progress to the user."""
	total = len(student_ids)
	deleted = 0
	failures = []

	try:
		for index, student_id in enumerate(student_ids, start=1):
			student_name = frappe.db.get_value("Student", student_id, "student_name") or student_id
			success, error = _delete_student_data(student_id)
			if success:
				deleted += 1
			else:
				failures.append(
					{"student": student_id, "student_name": student_name, "error": error}
				)

			frappe.publish_realtime(
				PROGRESS_EVENT,
				{
					"current": index,
					"total": total,
					"deleted": deleted,
					"failed": len(failures),
					"student": student_id,
					"student_name": student_name,
				},
				user=user,
			)

			# Commit in batches so progress is durable and locks stay short
			if index % 10 == 0:
				frappe.db.commit()

		frappe.db.commit()
	finally:
		frappe.cache().delete_value(RUNNING_CACHE_KEY)
		frappe.publish_realtime(
			DONE_EVENT,
			{"total": total, "deleted": deleted, "failures": failures},
			user=user,
		)


@frappe.whitelist()
def is_delete_running():
	"""Return whether a background deletion job is currently running."""
	_check_permission()
	return {"running": bool(frappe.cache().get_value(RUNNING_CACHE_KEY))}


# ---------------------------------------------------------------------------
# Export
# ---------------------------------------------------------------------------


@frappe.whitelist()
def export_students(filters=None):
	"""Stream an XLSX file of the matching students and their data."""
	_check_permission()
	from frappe.utils.xlsxutils import build_xlsx_response

	filters = _parse_filters(filters)
	where_clause, values = _build_conditions(filters)

	students = frappe.db.sql(
		f"""
		SELECT
			s.name AS student_id,
			s.student_name,
			CONCAT_WS(' ', s.first_name, s.middle_name, s.last_name) AS full_name,
			s.student_email_id AS email,
			s.student_mobile_number AS mobile_no,
			s.user,
			s.joining_date,
			s.educational_year,
			s.academic_year,
			(SELECT GROUP_CONCAT(DISTINCT pe.program SEPARATOR ', ')
				FROM `tabProgram Enrollment` pe WHERE pe.student = s.name) AS program,
			s.nationality,
			s.gender,
			s.date_of_birth,
			s.creation
		FROM `tabStudent` s
		WHERE {where_clause}
		ORDER BY s.creation DESC
		""",
		values,
		as_dict=True,
	)

	headers = [
		_("Student ID"),
		_("Student Name"),
		_("Full Name"),
		_("Email"),
		_("Mobile No"),
		_("User"),
		_("Joining Date"),
		_("Educational Year"),
		_("Academic Year"),
		_("Program"),
		_("Nationality"),
		_("Gender"),
		_("Date of Birth"),
		_("Registered On"),
	]

	rows = [headers]
	for s in students:
		rows.append(
			[
				s.student_id,
				s.student_name,
				s.full_name,
				s.email,
				s.mobile_no,
				s.user,
				cstr(s.joining_date),
				s.educational_year,
				s.academic_year,
				s.program,
				s.nationality,
				s.gender,
				cstr(s.date_of_birth),
				cstr(s.creation),
			]
		)

	build_xlsx_response(rows, "New Registered Students")
